File: C/src/0128_longest-consecutive-sequence.c
//
/* Created by rock on 6/6/20.
*/
#include "leetcode_functions.h"
int cmp(const void*a, const void*b)
{
    return (long long)*(int*)a - (long long)*(int*)b;
}

int longestConsecutive(int* a, int n)
{
    if (a == NULL || n == 0) {
        return 0;
    }
    qsort(a, n, sizeof(int), cmp);
    int t =a[0], cnt = 1, max = 1;
    for (int i = 1; i < n; i++) {
        if (a[i] == t) {
            continue;
        } else if ((long long)a[i] - t == 1L) {
            t = a[i];
            cnt++;
            if (max < cnt) {
                max = cnt;
            }
        } else {
            t = a[i];
            cnt = 1;
        }
    }
    return max;
}

File: python/_0001_0500/0128_longest-consecutive-sequence.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Author        : Rock Wayne 
# @Created       : 2020-05-06 08:00:00
# @Last Modified : 2020-05-06 08:00:00
# @Mail          : lostlorder@gmail.com
# @Version       : alpha-1.0

"""
# 给定一个未排序的整数数组，找出最长连续序列的长度。 
# 
#  要求算法的时间复杂度为 O(n)。 
# 
#  示例: 
# 
#  输入: [100, 4, 200, 1, 3, 2]
# 输出: 4
# 解释: 最长连续序列是 [1, 2, 3, 4]。它的长度为 4。 
#  Related Topics 并查集 数组

"""
from typing import List

import pytest


# leetcode submit region begin(Prohibit modification and deletion)

class Solution:

    def longestConsecutive(self, nums: List[int]) -> int:
        longest_streak = 0
        num_set = set(nums)

        for num in num_set:
            if num - 1 not in num_set:
                current_num = num
                current_streak = 1
                while current_num + 1 in num_set:
                    current_num += 1
                    current_streak += 1
                longest_streak = max(longest_streak, current_streak)
        return longest_streak


# leetcode submit region end(Prohibit modification and deletion)


@pytest.mark.parametrize("args,expected", [
    ([100, 4, 200, 1, 3, 2], 4),
])
def test_solutions(args, expected):
    assert Solution().longestConsecutive(args) == expected


if __name__ == '__main__':
    pytest.main(["-q", "--color=yes", "--capture=no", __file__])
